File: solve/lua/heap.lua
require "common" 

function swiftdown(top, last)
    local left = top*2
    if left <= last then
        local right = left + 1
        if right <= last and a[right] > a[left] then 
            left = right 
        end 
        if a[left] > a[top] then
            swap(top, left)
            swiftdown(left, last)
        end 
    end 
end

skeleton(function ()
    for i = math.floor(n/2), 1, -1 do
        swiftdown(i, n)
    end 
    for i = n, 2, -1 do
        swap(1, i)
        swiftdown(1, i - 1)
    end
end)

File: solve/cpp/heap.cc
#include "common.hh"

void swiftdown(int k, int last) {
    int i = k*2 + 1;
    if (i > last) return;
    if (i + 1 <= last && a[i + 1] > a[i]) ++i;
    if (a[i] > a[k]) {
        swapi(i, k);
        swiftdown(i, last);
    }
} 

void heapsort() {
    int last = n - 1;
    for (int i = n/2; i >= 0; --i) {
        swiftdown(i, last);
    }
    while (last != 0) {
        swapi(0, last);
        --last;
        swiftdown(0, last);
    } 
}

int main(int argc, char *argv[]) {
    if (init(argc, argv) == 0) {
        heapsort();
        output();
        return 0;
    }
    return 1;
}
